File: rust/src/bin/aoc_2022_13.rs
// Advent of Code, 2022 day 13

use aoc_rust::stdin_lines;
use std::cmp::Ordering;

/// Chunks just iterate over the input strings (collecting digits into integers, dropping commas)
#[derive(Clone, Copy, Debug, PartialEq)]
enum Chunk {
    Open,
    Close,
    Number(u32),
}

struct ChunkIterator<'a> {
    chars: std::iter::Peekable<std::str::Chars<'a>>, // peekable char iterator
    closes_due: u32,                                 // number of close parens expected
    // next two fields support promotion (e.g., of 7 into [7])
    next_chunk: Option<Chunk>, // next chunk to be returned (instead of following index)
    last: Option<u32>,         // previous value (from promotion)
}

impl<'a> ChunkIterator<'a> {
    fn new(s: &str) -> ChunkIterator {
        ChunkIterator {
            chars: s.chars().peekable(),
            last: None,
            next_chunk: None,
            closes_due: 0,
        }
    }
    fn promote(&mut self) {
        self.next_chunk = Some(Chunk::Number(self.last.unwrap()));
        self.closes_due += 1;
    }
}

impl<'a> Iterator for ChunkIterator<'a> {
    type Item = Chunk;

    fn next(&mut self) -> Option<Self::Item> {
        // deal with next_chunk if it has been set
        if let Some(next_chunk) = self.next_chunk {
            match next_chunk {
                Chunk::Close => {
                    if self.closes_due == 1 {
                        self.next_chunk = None;
                    } else {
                        self.closes_due -= 1;
                    }
                }
                Chunk::Number(_) => {
                    self.next_chunk = Some(Chunk::Close);
                }
                Chunk::Open => panic!("Unexpected"),
            }
            return Some(next_chunk);
        }

        let mut c = self.chars.next()?;

        // skip a comma
        if c == ',' {
            c = self.chars.next().unwrap();
        }

        // Handle open
        if c == '[' {
            self.last = None;
            return Some(Chunk::Open);
        }

        // Handle close
        if c == ']' {
            self.last = None;
            return Some(Chunk::Close);
        }

        // Handle numbers
        assert!(c.is_ascii_digit());
        let mut number_string: String = String::new();
        number_string.push(c);
        while let Some(cp) = self.chars.peek() {
            if cp.is_ascii_digit() {
                number_string.push(*cp);
                self.chars.next();
            } else {
                break;
            }
        }
        let number: u32 = number_string.parse().unwrap();
        self.last = Some(number);
        Some(Chunk::Number(number))
    }
}

fn compare(a: &str, b: &str) -> Ordering {
    let mut i = ChunkIterator::new(a);
    let mut j = ChunkIterator::new(b);
    let mut open_lists: u32 = 0;
    loop {
        match (i.next().unwrap(), j.next().unwrap()) {
            (Chunk::Open, Chunk::Open) => open_lists += 1,
            (Chunk::Open, Chunk::Close) => return Ordering::Greater,
            (Chunk::Open, Chunk::Number(_)) => j.promote(),
            (Chunk::Close, Chunk::Close) => {
                if open_lists == 1 {
                    return Ordering::Equal;
                } else {
                    open_lists += 1;
                }
            }
            (Chunk::Close, _) => return Ordering::Less,
            (Chunk::Number(_), Chunk::Open) => i.promote(),
            (Chunk::Number(_), Chunk::Close) => return Ordering::Greater,
            (Chunk::Number(p), Chunk::Number(q)) => {
                if p != q {
                    return p.cmp(&q);
                }
            }
        }
    }
}

fn sum_ordered(lines: &[String]) -> u32 {
    let mut pair: u32 = 1;
    let mut sum: u32 = 0;
    let mut lines_iter = lines.iter();
    while let Some(a) = lines_iter.next() {
        let b = lines_iter.next().unwrap();
        let blank = lines_iter.next();
        assert!(blank.map_or_else(|| true, |s| s.trim().is_empty()));
        if compare(a, b) != Ordering::Greater {
            sum += pair;
        }
        pair += 1;
    }
    sum
}

fn divider_positions(lines: &[String]) -> usize {
    // Take all non-blank packets
    let mut packets: Vec<&String> = lines.iter().filter(|s| !s.trim().is_empty()).collect();

    // Add two divider packets
    let divider1 = String::from("[[2]]");
    let divider2 = String::from("[[6]]");
    packets.push(&divider1);
    packets.push(&divider2);

    packets.sort_by(|x, y| compare(x, y));

    packets
        .iter()
        .enumerate()
        .filter(|(_, &s)| s == &divider1 || s == &divider2)
        .map(|(i, _)| i + 1)
        .product()
}

fn main() {
    let lines: Vec<String> = stdin_lines().collect();

    println!("{}", sum_ordered(&lines));
    println!("{}", divider_positions(&lines));
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_chunk() {
        let mut chunk_iter = ChunkIterator::new("[1,[234],56]");
        assert_eq!(chunk_iter.next(), Some(Chunk::Open));
        assert_eq!(chunk_iter.next(), Some(Chunk::Number(1)));
        assert_eq!(chunk_iter.next(), Some(Chunk::Open));
        assert_eq!(chunk_iter.next(), Some(Chunk::Number(234)));
        assert_eq!(chunk_iter.next(), Some(Chunk::Close));
        assert_eq!(chunk_iter.next(), Some(Chunk::Number(56)));
        assert_eq!(chunk_iter.next(), Some(Chunk::Close));
        assert_eq!(chunk_iter.next(), None);
    }

    #[test]
    fn test_compare() {
        assert_eq!(compare("[1,1,3,1,1]", "[1,1,5,1,1]"), Ordering::Less);
        assert_eq!(compare("[[1],[2,3,4]]", "[[1],4]"), Ordering::Less);
        assert_eq!(compare("[9]", "[[8,7,6]]"), Ordering::Greater);
        assert_eq!(compare("[[4,4],4,4]", "[[4,4],4,4,4]"), Ordering::Less);
        assert_eq!(compare("[7,7,7,7]", "[7,7,7]"), Ordering::Greater);
        assert_eq!(compare("[]", "[3]"), Ordering::Less);
        assert_eq!(compare("[[[]]]", "[[]]"), Ordering::Greater);
        assert_eq!(
            compare("[1,[2,[3,[4,[5,6,7]]]],8,9]", "[1,[2,[3,[4,[5,6,0]]]],8,9]"),
            Ordering::Greater
        );
        assert_eq!(
            compare(
                "[[8,[[2,6,0,9],[4,9,5,5,3],[8],8,3],[1]],[0,[7,[8,8,8,1]],[],[[0],[8,6,9,9,1],2,6,2]],[],[1,[2,1,[1],8],[],8],[[9,7],[7,10],[[],9,9,7,3],[2,[],6],[[],5,[7,6,7,6,0],[3],[3,10,10,10]]]]", 
                "[[[[8],1,[1,1,3,5,1],[0,3]],[],4,0,5],[],[3,3,[9,4,5,[4,2,3]],7,9],[7],[[[2,1]],[[2,9,10,0,6],[3,0,8],[0,3,2,9,1]]]]"
            ),
            Ordering::Less
        );
    }
}

File: others/zig/src/aoc_2022_13.zig
// Advent of Code, 2022 day 13

const std = @import("std");
const rw = @import("rw");

const ChunkTag = enum { open, close, number };

const Chunk = union(ChunkTag) {
    open: void,
    close: void,
    number: usize,
};

const ChunkIterator = struct {
    buffer: []const u8,
    index: usize,
    last: ?usize,
    next_chunk: ?Chunk,
    closes_due: usize,

    const Self = @This();

    fn next(self: *Self) !?Chunk {
        if (self.next_chunk) |next_value| {
            if (@as(ChunkTag, next_value) == ChunkTag.close) {
                if (self.closes_due == 1) {
                    self.next_chunk = null;
                } else {
                    self.closes_due -= 1;
                }
            } else {
                std.debug.assert(@as(ChunkTag, next_value) == ChunkTag.number);
                self.next_chunk = Chunk{ .close = {} };
            }
            return next_value;
        }
        if (self.index < self.buffer.len and self.buffer[self.index] == ',') {
            self.index += 1;
        }
        if (self.index >= self.buffer.len) {
            return null;
        }
        if (self.buffer[self.index] == '[') {
            self.index += 1;
            self.last = null;
            return Chunk{ .open = {} };
        }
        if (self.buffer[self.index] == ']') {
            self.index += 1;
            self.last = null;
            return Chunk{ .close = {} };
        }
        var number_of_digits: usize = 0;
        while (std.ascii.isDigit(self.buffer[self.index + number_of_digits])) {
            number_of_digits += 1;
        }
        std.debug.assert(number_of_digits >= 1);
        const value = try std.fmt.parseUnsigned(usize, self.buffer[self.index .. self.index + number_of_digits], 10);
        self.index += number_of_digits;
        self.last = value;
        return Chunk{ .number = value };
    }

    fn promote(self: *Self) void {
        self.next_chunk = Chunk{ .number = self.last orelse unreachable };
        self.closes_due += 1;
    }
};

// Return iterator over list chunks
fn chunk(s: []const u8) ChunkIterator {
    return ChunkIterator{
        .buffer = s,
        .index = 0,
        .last = null,
        .next_chunk = null,
        .closes_due = 0,
    };
}

test "chunk" {
    var iter = chunk("[1,[234],56]");
    try std.testing.expectEqualDeep(Chunk{ .open = {} }, (try iter.next()).?);
    try std.testing.expectEqualDeep(Chunk{ .number = 1 }, (try iter.next()).?);
    try std.testing.expectEqualDeep(Chunk{ .open = {} }, (try iter.next()).?);
    try std.testing.expectEqualDeep(Chunk{ .number = 234 }, (try iter.next()).?);
    try std.testing.expectEqualDeep(Chunk{ .close = {} }, (try iter.next()).?);
    try std.testing.expectEqualDeep(Chunk{ .number = 56 }, (try iter.next()).?);
    try std.testing.expectEqualDeep(Chunk{ .close = {} }, (try iter.next()).?);
    try std.testing.expect(try iter.next() == null);
}

const Comparison = enum {
    less,
    equal,
    greater,
};

const ComparisonError = error{ UnexpectedEnd, Malformed };

fn compare(a: []const u8, b: []const u8) !Comparison {
    var i: ChunkIterator = chunk(a);
    var j: ChunkIterator = chunk(b);
    var open_lists: usize = 0;

    while (true) {
        var x: Chunk = (try i.next()) orelse return ComparisonError.UnexpectedEnd;
        var y: Chunk = (try j.next()) orelse return ComparisonError.UnexpectedEnd;
        switch (x) {
            ChunkTag.open => {
                switch (y) {
                    ChunkTag.open => open_lists += 1,
                    ChunkTag.close => return Comparison.greater,
                    ChunkTag.number => j.promote(),
                }
            },
            ChunkTag.close => {
                switch (y) {
                    ChunkTag.open => return Comparison.less,
                    ChunkTag.close => if (open_lists == 1) {
                        return Comparison.equal;
                    } else {
                        open_lists -= 1;
                    },
                    ChunkTag.number => return Comparison.less,
                }
            },
            ChunkTag.number => |x_value| {
                switch (y) {
                    ChunkTag.open => i.promote(),
                    ChunkTag.close => return Comparison.greater,
                    ChunkTag.number => |y_value| {
                        if (x_value < y_value) {
                            return Comparison.less;
                        }
                        if (x_value > y_value) {
                            return Comparison.greater;
                        }
                    },
                }
            },
        }
    }
}

test "compare" {
    try std.testing.expectEqual(
        Comparison.less,
        try compare(
            "[1,1,3,1,1]",
            "[1,1,5,1,1]",
        ),
    );

    try std.testing.expectEqual(
        Comparison.less,
        try compare(
            "[[1],[2,3,4]]",
            "[[1],4]",
        ),
    );

    try std.testing.expectEqual(
        Comparison.greater,
        try compare(
            "[9]",
            "[[8,7,6]]",
        ),
    );

    try std.testing.expectEqual(
        Comparison.less,
        try compare(
            "[[4,4],4,4]",
            "[[4,4],4,4,4]",
        ),
    );

    try std.testing.expectEqual(
        Comparison.greater,
        try compare(
            "[7,7,7,7]",
            "[7,7,7]",
        ),
    );

    try std.testing.expectEqual(
        Comparison.less,
        try compare(
            "[]",
            "[3]",
        ),
    );

    try std.testing.expectEqual(
        Comparison.greater,
        try compare(
            "[[[]]]",
            "[[]]",
        ),
    );

    try std.testing.expectEqual(
        Comparison.greater,
        try compare(
            "[1,[2,[3,[4,[5,6,7]]]],8,9]",
            "[1,[2,[3,[4,[5,6,0]]]],8,9]",
        ),
    );
}

test "compare_extra" {
    try std.testing.expectEqual(
        Comparison.less,
        try compare(
            "[[8,[[2,6,0,9],[4,9,5,5,3],[8],8,3],[1]],[0,[7,[8,8,8,1]],[],[[0],[8,6,9,9,1],2,6,2]],[],[1,[2,1,[1],8],[],8],[[9,7],[7,10],[[],9,9,7,3],[2,[],6],[[],5,[7,6,7,6,0],[3],[3,10,10,10]]]]",
            "[[[[8],1,[1,1,3,5,1],[0,3]],[],4,0,5],[],[3,3,[9,4,5,[4,2,3]],7,9],[7],[[[2,1]],[[2,9,10,0,6],[3,0,8],[0,3,2,9,1]]]]",
        ),
    );
}

fn sum_ordered(lines: [][]const u8) !usize {
    var pair: usize = 1;
    var sum: usize = 0;
    while (pair * 3 - 2 < lines.len) : (pair += 1) {
        const result = try compare(lines[pair * 3 - 3], lines[pair * 3 - 2]);

        if (result != Comparison.greater) {
            sum += pair;
        }
    }
    return sum;
}

fn less_than(_: void, a: []const u8, b: []const u8) bool {
    const result = compare(a, b) catch return false;
    return result == Comparison.less;
}

fn divider_positions(allocator: std.mem.Allocator, lines: [][]const u8) !usize {

    // Copy only non-blank lines
    var packets = std.ArrayList([]const u8).init(allocator);
    defer packets.deinit();
    for (lines) |line| {
        if (line.len > 0) {
            try packets.append(line);
        }
    }

    // Add dividers
    const divider1 = "[[2]]";
    const divider2 = "[[6]]";
    try packets.append(divider1);
    try packets.append(divider2);

    std.sort.insertion([]const u8, packets.items, {}, less_than);

    var product: usize = 1;
    for (packets.items, 1..) |packet, i| {
        if (std.mem.eql(u8, packet, divider1) or std.mem.eql(u8, packet, divider2)) {
            product *= i;
        }
    }
    return product;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read text input lines
    const lines: [][]const u8 = try rw.read_lines(allocator);
    defer {
        for (lines) |line| {
            allocator.free(line);
        }
        allocator.free(lines);
    }

    try rw.write_int(usize, try sum_ordered(lines));
    try rw.write_int(usize, try divider_positions(allocator, lines));
}
